File: Data Structures & Algorithms/search-for-word/submission-4.go
func helper(board [][]byte, word string, k int, row int, col int,n int ,m int) bool {
    if k == len(word) {
        return true
    }
    if k > len(word) {
        return false
    }

    dirs := [][] int {{0,1},{1,0},{-1,0},{0,-1}}
    res :=false
    for _,dir := range(dirs) {
        r:=row+dir[0]
        c:=col+dir[1]
        
        if r<n && c< m && r>=0 && c>=0 && board[r][c]!='#' && board[r][c]==word[k] {
            val := board[r][c]
            board[r][c]='#'
            res=res || helper(board,word,k+1,r,c,n,m)
            board[r][c]=val
        }
    }
    return res
}

func exist(board [][]byte, word string) bool {
    n:=len(board)
    m:=len(board[0])

    for i:=0; i<n; i++ {
        for j:=0; j<m; j++ {
            
            if word[0]==board[i][j] {
                val :=board[i][j]
                board[i][j]='#'
                res:=helper(board,word,1,i,j,n,m)
                if res {
                    return res
                }
                board[i][j]=val
            }
        }
    }
    return false
}

File: Data Structures & Algorithms/search-for-word-ii/submission-4.py
class TrieNode:
    def __init__(self):
        self.child = {}
        self.isWord = False
    def addWord(self,word):
        curr=self
        for c in word:
            if c not in curr.child:
                curr.child[c]=TrieNode()
            curr=curr.child[c]
        curr.isWord=True


class Solution:

    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        row=len(board)
        col=len(board[0])
        root = TrieNode()
        for word in words:
            root.addWord(word)

        res=set()
        vis=set()
        def dfs(r,c,node,word):
            if r>=row or r<0 or c>=col or c<0 or (r,c) in vis or board[r][c] not in node.child:
                return
            vis.add((r,c))
            node = node.child[board[r][c]]
            word+=board[r][c]
            if node.isWord:
                res.add(word)
            
            dfs(r+1,c,node,word)
            dfs(r-1,c,node,word)
            dfs(r,c+1,node,word)
            dfs(r,c-1,node,word)
            vis.remove((r,c))

        for r in range(len(board)):
            for c in range(len(board[0])):
                dfs(r,c,root,'')
        return list(res)


        